Implementation of buffer blit #2905

SnapperTT opened the pull request:
Addresses #2905 

Cubes example has been modified to test copying buffers. When its ready for merging these changes will be reverted.

Only the OpenGL backend is supported for now.

Adds a new cap `BGFX_CAPS_BUFFER_BLIT`. Copies are encoded by calling `bgfx::blit`. The parameters for the copy are packed into the `BlitInfo` struct and unpacked by the backend. A buffer blit is detected when both the src and dest texture are `BGFX_INVALID_HANDLE` and the depth value is set to `BGFX_BUFFER_BLIT_MAGIC`.

Further work to be done: 
- Implement copies for different buffer types (Dynamic, Index, Indirect, etc)
- Implement dx11 copy

bkaradzic commented:
This encoding buffer into texture is totally jenky... See `RenderItem`...

There are two ideas...

1. `BlitItem` should be renamed to `BlitTextureItem`, there should be new struct called `BlitBufferItem`, and then `BlitItem` becomes:

```cpp
	union BlitItem
	{
		BlitTextureItem texture;
		BlitBufferItem  buffer;
	};
```

And the only thing left is to add some flag to know which one is encoded into `BlitItem`.


2. Another thing you can do is to use `Handle` instead of `TextureHandle` since bgfx internally can store handle type inside `Handle`. Could be useful if it's possible to blit between buffer and texture.

SnapperTT commented:
I've converted `BlitItem` into a union. You expressed interest in being able to blit buffer <-> texture so I changed the buffer blit api to work with *bytes* and not *elements/vertices*.

Eg:
```
bgfx::blit(0, m_vbh, m_tempVbh, 4*sizeof(PosColorVertex), 4*sizeof(PosColorVertex));
```

bkaradzic commented:
This feature needs to be well thought out... Currently it lacks multiple backend implementations. Also multiple different directions of blitting from buffer -> texture, texture -> buffer, etc.

Buffer -> Buffer
https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400767-copyfrombuffer

Buffer -> Texture
https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400771-copyfrombuffer

Texture -> Buffer
https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400756-copyfromtexture

bkaradzic commented:
I improved internal handle, so it can be used to detect if resource is buffer or texture:
https://github.com/bkaradzic/bgfx/commit/8a405a0c73f4f628056fafe6bd79504a01492d47
